hstream/src/HStream/Server/Handler.hs

```diff
@@ -361,165 +361,194 @@ sendToClient zkHandle qid sc@SourceConnector {..} ss@streamSend = do
-      returnResp $ SubscribeResponse subscribeRequestSubscriptionId
-
-    doClean :: IO ()
-    doClean = atomically $ updateReaderStatus subscribedReaders Released subscribeRequestSubscriptionId
+    getSubscriptionOffset :: SubscriptionId -> IO SubscriptionOffset
+    getSubscriptionOffset subscriptionId = atomically $ do
+      store <- readTVar subscriptions
+      let Subscription{..} = fromJust $ HM.lookup subscriptionId store
+      return $ fromJust subscriptionOffset
+
+    getStartLSN :: SubscriptionOffset -> S.LSN
+    getStartLSN SubscriptionOffset{..} = case fromJust subscriptionOffsetOffset of
+      SubscriptionOffsetOffsetSpecialOffset _ -> error "shoud not reach here"
+      SubscriptionOffsetOffsetRecordOffset RecordId{..} -> recordIdBatchId
+
+getStreamName :: SubscriptionId -> TVar (HM.HashMap SubscriptionId Subscription)-> IO TL.Text
+getStreamName subscriptionId tStore = atomically $ do
+  store <- readTVar tStore
+  let Subscription{..} = fromJust $ HM.lookup subscriptionId store
+  return subscriptionStreamName
 
 deleteSubscriptionHandler
   :: ServerContext
   -> ServerRequest 'Normal DeleteSubscriptionRequest Empty
   -> IO (ServerResponse 'Normal Empty)
 deleteSubscriptionHandler ServerContext{..} (ServerNormalRequest _metadata req@DeleteSubscriptionRequest{..}) = defaultExceptionHandle $ do
   Log.debug $ "Receive deleteSubscription request: " <> Log.buildString (show req)
-  (reader, Subscription{..}) <- lookupSubscribedReaders subscribedReaders deleteSubscriptionRequestSubscriptionId
-  let streamName = transToStreamName $ TL.toStrict subscriptionStreamName
-  isExist <- S.doesStreamExists scLDClient streamName
-  when isExist $ do
-      logId <- S.getUnderlyingLogId scLDClient streamName
-      ckpReaderStopReading reader logId
-  atomically $ do
-    deleteSubscribedReaders subscribedReaders subscriptionSubscriptionId
-    modifyTVar' subscribeHeap $ Map.delete subscriptionSubscriptionId
+
+  mSubscription <- atomically $ do
+    store <- readTVar subscriptions
+    case HM.lookup deleteSubscriptionRequestSubscriptionId store of
+      Just sub -> do
+        let newStore = HM.delete deleteSubscriptionRequestSubscriptionId store
+        writeTVar subscriptions newStore
+        return (Just sub)
+      Nothing -> return Nothing
+
+  case mSubscription of
+    Just Subscription{..} -> do
+      mRes <- atomically $ do
+        store <- readTVar subscribeRuntimeInfo
+        case HM.lookup deleteSubscriptionRequestSubscriptionId store of
+          Just ldreader -> do
+            let newStore = HM.delete deleteSubscriptionRequestSubscriptionId store
+            writeTVar subscribeRuntimeInfo newStore
+            return (Just ldreader)
+          Nothing -> return Nothing
+
+      case mRes of
+        Just ldreader -> do
+          -- stop ldreader
+          logId <- S.getUnderlyingLogId scLDClient (transToStreamName $ TL.toStrict subscriptionStreamName)
```

Comment:
I think we should execute `doesStreamExists` before this `getUnderlyingLogId`, because if the stream is deleted (maybe by some mistake operation or some unknown reasons), `getUnderlyingLogId` will throw an exception, and the final result is: we delete the subscription, but also receive an exception at the same time. Do an additional check can ignore this case.